### PR TITLE
Require all profiles to be specified when using multiple filters

### PR DESCRIPTION
**Pull request by chinyitan**

### CODE REVIEW

File: io_directory_example/settings/lens_system3_config.yml [hunk 1]
```diff
@@ -17,9 +17,14 @@ model:
   lens_light:
     - 'SERSIC_ELLIPSE'
     - 'SERSIC_ELLIPSE'
+    - 'SERSIC_ELLIPSE'
+    - 'SERSIC_ELLIPSE'
+  lens_light_band_index: [0,0,1,1]
```
Comment (ajshajib) [inline io_directory_example/settings/lens_system3_config.yml:22]: Add spaces after comma.

File: io_directory_example/settings/lens_system3_config.yml [hunk 2]
```diff
@@ -17,9 +17,14 @@ model:
...
 
   source_light:
     - 'SHAPELETS'
+    - 'SHAPELETS'
+  source_light_band_index: [0,1]
```
Comment (ajshajib) [inline io_directory_example/settings/lens_system3_config.yml:27]: Add space after comma.

File: io_directory_example/settings/lens_system3_config.yml (hunk 1 above)
Comment (ajshajib) [reply io_directory_example/settings/lens_system3_config.yml:22]: Change `lens_light_band_index` to `lens_light_band_indices`, then it's clearer that this value has to be something like a `list`, and not just a single `float`.

File: io_directory_example/settings/lens_system3_config.yml (hunk 2 above)
Comment (ajshajib) [reply io_directory_example/settings/lens_system3_config.yml:27]: Change `source_light_band_index` to `source_light_band_indices`.

File: test/test_processor/test_config.py [hunk 3]
```diff
@@ -428,6 +428,11 @@ def test_fill_in_fixed_from_settings(self):
                                                         fixed)
         assert fixed == [{'n_sersic': 4.}]
 
+        fixed2 = [{}, {}, {}, {}]
```
Comment (ajshajib) [inline test/test_processor/test_config.py:431]: This line is redundant.

File: dolphin/processor/config.py [hunk 4]
```diff
@@ -490,15 +482,24 @@ def get_index_lens_light_model_list(self):
...
+                    index_list_settings = \
+                       self.settings['model']['lens_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
+                    for i, model in enumerate(index_list_settings):
+                        index_list[model].append(i)
+                    for k in index_list:
+                        assert k != [], "One of the bands have no lens light"
+                    return index_list
+                else:
+                    raise ValueError('Missing lens_light_band_index')
```
Comment (ajshajib) [inline dolphin/processor/config.py:502]: Add more description to the error message as `'Missing "lens_light_band_indices" in the settings file!'`

File: dolphin/processor/config.py [hunk 5]
```diff
@@ -508,17 +509,25 @@ def get_index_source_light_model_list(self):
...
+                    index_list_settings =\
+                        self.settings['model']['source_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
+                    for i, model in enumerate(index_list_settings):
+                        index_list[model].append(i)
+                    for k in index_list:
+                        assert k != [], "One of the bands have no source light"
+                    return index_list
+                else:
+                    raise ValueError('Missing source_light_band_index')
```
Comment (ajshajib) [inline dolphin/processor/config.py:530]: Add more description to the error message as `'Missing "source_light_band_indices" in the settings file!'`

File: dolphin/processor/config.py [hunk 6]
```diff
@@ -508,17 +509,25 @@ def get_index_source_light_model_list(self):
...
+                index_list = [[]]
+                for k, model in enumerate(
+                        self.settings['model']['source_light']):
+                    index_list[0].append(k)
+                return index_list
+            else:
+                if 'source_light_band_index' in self.settings['model']:
+                    index_list_settings =\
+                        self.settings['model']['source_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
```
Comment (ajshajib) [inline dolphin/processor/config.py:523]: You can simplify this line as:

```
index_list = [[]] * self.band_number
```

File: dolphin/processor/config.py [hunk 7]
```diff
@@ -490,15 +482,24 @@ def get_index_lens_light_model_list(self):
...
+                index_list = [[]]
+                for k, model in enumerate(
+                        self.settings['model']['lens_light']):
+                    index_list[0].append(k)
+                return index_list
+            else:
+                if 'lens_light_band_index' in self.settings['model']:
+                    index_list_settings = \
+                       self.settings['model']['lens_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
```
Comment (ajshajib) [inline dolphin/processor/config.py:495]: You can simplify this line as:

```
index_list = [[]] * self.band_number
```

File: dolphin/processor/config.py [hunk 8]
```diff
@@ -490,15 +482,24 @@ def get_index_lens_light_model_list(self):
...
-            return index_lens_light_model_list
+            if self.band_number == 1:
+                index_list = [[]]
+                for k, model in enumerate(
+                        self.settings['model']['lens_light']):
+                    index_list[0].append(k)
+                return index_list
+            else:
+                if 'lens_light_band_index' in self.settings['model']:
+                    index_list_settings = \
```
Comment (ajshajib) [inline dolphin/processor/config.py:493]: It is unnecessary to create a new variable here. You can simply use `self.settings['model']['lens_light_band_index']` for that one time you need it in line 496.

File: dolphin/processor/config.py [hunk 9]
```diff
@@ -508,17 +509,25 @@ def get_index_source_light_model_list(self):
...
+        if 'source_light' in self.settings['model']:
+            if self.band_number == 1:
+                index_list = [[]]
+                for k, model in enumerate(
+                        self.settings['model']['source_light']):
+                    index_list[0].append(k)
+                return index_list
+            else:
+                if 'source_light_band_index' in self.settings['model']:
+                    index_list_settings =\
```
Comment (ajshajib) [inline dolphin/processor/config.py:521]: It's unnecessary to create this new variable `index_list_settings`.

File: dolphin/processor/config.py [hunk 10]
```diff
@@ -490,15 +482,24 @@ def get_index_lens_light_model_list(self):
...
+                return index_list
+            else:
+                if 'lens_light_band_index' in self.settings['model']:
+                    index_list_settings = \
+                       self.settings['model']['lens_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
+                    for i, model in enumerate(index_list_settings):
+                        index_list[model].append(i)
+                    for k in index_list:
+                        assert k != [], "One of the bands have no lens light"
```
Comment (ajshajib) [inline dolphin/processor/config.py:499]: Add a punctuation (. or !) at the end of the error message.

File: dolphin/processor/config.py [hunk 11]
```diff
@@ -508,17 +509,25 @@ def get_index_source_light_model_list(self):
...
+                return index_list
+            else:
+                if 'source_light_band_index' in self.settings['model']:
+                    index_list_settings =\
+                        self.settings['model']['source_light_band_index']
+                    index_list = [[] for _ in range(self.band_number)]
+                    for i, model in enumerate(index_list_settings):
+                        index_list[model].append(i)
+                    for k in index_list:
+                        assert k != [], "One of the bands have no source light"
```
Comment (ajshajib) [inline dolphin/processor/config.py:527]: Add a punctuation (. or !) at the end of the error message.

File: test/test_processor/test_config.py (hunk 3 above)
Comment (chinyitan) [reply test/test_processor/test_config.py:431]: Do you mean just line 431 or the entire fixed2 section (line 431 to 434)? I am quite sure line 431 is necessary if we decide to  add line 432 to 434.

File: dolphin/processor/config.py (hunk 6 above)
Comment (chinyitan) [reply dolphin/processor/config.py:521]: That was what I first tried too, but I think that will result in the list containing the same list object N times, i.e. if we add an index in the first list, the second list will also have the same index.

Comment (ajshajib) [reply dolphin/processor/config.py:521]: O I see, didn't realize that would happen, then the current way is the right way!

File: test/test_processor/test_config.py (hunk 3 above)
Comment (ajshajib) [reply test/test_processor/test_config.py]: O I see that `self.config3.fill_in_fixed_from_settings()` requires `fixed2` as  an argument, which I missed before. This is fine.

File: io_directory_example/settings/lens_system3_config.yml [hunk 12]
```diff
@@ -17,9 +17,14 @@ model:
   lens_light:
     - 'SERSIC_ELLIPSE'
     - 'SERSIC_ELLIPSE'
+    - 'SERSIC_ELLIPSE'
+    - 'SERSIC_ELLIPSE'
+  lens_light_band_indices: [0,0,1,1]
```
Comment (ajshajib) [inline io_directory_example/settings/lens_system3_config.yml:22]: Add spaces after the commas.

File: io_directory_example/settings/lens_system3_config.yml [hunk 13]
```diff
@@ -17,9 +17,14 @@ model:
...
 
   source_light:
     - 'SHAPELETS'
+    - 'SHAPELETS'
+  source_light_band_indices: [0,1]
```
Comment (ajshajib) [inline io_directory_example/settings/lens_system3_config.yml:27]: Add a space after the comma.

**Review by ajshajib — APPROVED** [review]

**Pull request merged**